pytorch_lightning/core/step_result.py

```diff
@@ -320,12 +319,16 @@ def get_epoch_log_metrics(self, add_dataloader_idx=False) -> dict:
             if options['logger'] and options['on_epoch']:
                 if isinstance(self[k], Metric):
                     result[dl_key] = self[k].compute().detach()
+                    self[k].reset()
                 else:
                     result[dl_key] = self[k]
 
             if k in self and not options['on_epoch'] and isinstance(self[k], Metric):
-                # compute metric on epoch anyway so state does not accumulate
+                # reset metric anyway so state does not accumulate
+                # NOTE: we must compute before reseting just in case the computed value is needed
+                # later (i.e. if the step metric gets visited first, and then the epoch metric)
                 self[k].compute()
+                self[k].reset()
```

Comment:
Why reset is being added there ? `self[k].compute()` and `self[k].reset()` makes `compute()` useless no ?

Reply:
`compute()` will save the computed value to `self._computed`, so if `compute()` is called again without any `update()` call, the cached value will be provided. We still call `reset()` so that we don't get too much memory usage by needlessly accumulating tensors over epochs.

Reply:
This is important if we have a metric that is logged on both step and epoch. Since it will show up twice in the results meta dict, we want to reset the step metric on epoch end, but still allow the epoch metric to be computed.